guidance/models/transformers/_transformers.py

```diff
@@ -86,6 +86,21 @@ def __init__(
                     )
                 byte_tokens[i] = byte_coded.replace(space_prefix, b" ")
 
+        elif hasattr(transformers_tokenizer, "get_vocab"):
+            vocab = transformers_tokenizer.get_vocab()
+            for i in range(len(transformers_tokenizer)):
+                if i in special_tokens_map:
+                    byte_coded = special_tokens_map[i].encode()
+                else:
+                    token = transformers_tokenizer.convert_ids_to_tokens(i)
+                    if isinstance(token, bytes):
+                        byte_coded = token
+                    elif isinstance(token, str):
+                        byte_coded = token.encode()
+                    else:
+                        raise ValueError(f"Unexpected token type: {type(token)}")
+                byte_tokens[i] = byte_coded
+
```

Comment:
I think this looks great! I was worried we'd have to start doing a call into `tiktoken` directly to build a full byte set. Thanks for the great contribution :)

Reply:
Ah, I may have spoke too soon. Based on the test failures, we may have to do some `tiktoken` magic after all to build the full token set :(

@ambisinister , are you able to do simple gen and select calls with phi-3-small from this PR?

Reply:
Hi, I initially got started on this PR to make the following work:

https://github.com/ambisinister/wilt/blob/main/models/phi_model.py

Which seems to work properly -- from the test results it looks like there may be some issue with the end of text tokens? I'll see if I can reproduce today

Reply:
```
import guidance
from guidance import gen, select
from guidance import models

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_obj = AutoModelForCausalLM.from_pretrained(
    "microsoft/Phi-3-small-8k-instruct",
    device_map="auto",
    torch_dtype="auto",
    load_in_8bit=True,
    load_in_4bit=False,
    trust_remote_code=True,
    use_flash_attention_2=True
)
tokenizer = AutoTokenizer.from_pretrained("microsoft/Phi-3-small-8k-instruct",
                                          use_fast=False,
                                          trust_remote_code=True)
model = models.Transformers(model_obj, tokenizer, echo=False)

@guidance
def select_test(lm):
    lm = lm + select(["Scott is a person", "Scott is a persimmon"])
    lm = lm + "-Hi there."
    return lm

@guidance
def gen_test(lm):
    lm = lm + "Scott is a" + gen(name="test", max_tokens=40) +"-Hi there"
    return lm

model += select_test()
model += gen_test()

print(str(model))
```

The output of the above is:

```
Scott is a person-Hi there.Scott is a person who loves to solve puzzles and riddles. He has a unique way of organizing his bookshelf. He arranges his books in a specific order based on the number of letters in the title of-Hi there
```

I did the above in a venv as well, below is pip list --local

```
(guidancevenv) ambi@Lain-Navi:~/code/guidance$ pip list --local
Package                  Version
------------------------ -----------
accelerate               0.32.1
annotated-types          0.7.0
anyio                    4.4.0
bitsandbytes             0.43.1
certifi                  2024.7.4
charset-normalizer       3.3.2
diskcache                5.6.3
distro                   1.9.0
einops                   0.8.0
exceptiongroup           1.2.1
filelock                 3.15.4
flash-attn               2.5.9.post1
fsspec                   2024.6.1
h11                      0.14.0
httpcore                 1.0.5
httpx                    0.27.0
huggingface-hub          0.23.4
idna                     3.7
iniconfig                2.0.0
Jinja2                   3.1.4
MarkupSafe               2.1.5
mpmath                   1.3.0
networkx                 3.3
numpy                    1.26.4
nvidia-cublas-cu12       12.1.3.1
nvidia-cuda-cupti-cu12   12.1.105
nvidia-cuda-nvrtc-cu12   12.1.105
nvidia-cuda-runtime-cu12 12.1.105
nvidia-cudnn-cu12        8.9.2.26
nvidia-cufft-cu12        11.0.2.54
nvidia-curand-cu12       10.3.2.106
nvidia-cusolver-cu12     11.4.5.107
nvidia-cusparse-cu12     12.1.0.106
nvidia-nccl-cu12         2.20.5
nvidia-nvjitlink-cu12    12.5.82
nvidia-nvtx-cu12         12.1.105
openai                   1.35.10
ordered-set              4.1.0
packaging                24.1
pip                      22.0.2
platformdirs             4.2.2
pluggy                   1.5.0
protobuf                 5.27.2
psutil                   6.0.0
pybind11                 2.13.1
pydantic                 2.8.2
pydantic_core            2.20.1
pytest                   8.2.2
PyYAML                   6.0.1
regex                    2024.5.15
requests                 2.32.3
safetensors              0.4.3
setuptools               59.6.0
sniffio                  1.3.1
sympy                    1.13.0
tiktoken                 0.7.0
tokenizers               0.19.1
tomli                    2.0.1
torch                    2.3.1
tqdm                     4.66.4
transformers             4.42.3
triton                   2.3.1
typing_extensions        4.12.2
urllib3                  2.2.2
wheel                    0.43.0
```

Reply:
It looks like this may be causing trouble with the Phi2 tests.